### PR TITLE
SW-4731 Logic for random on-demand plot selection

### PR DESCRIPTION
**Pull request by sgrimm**
In preparation for creating monitoring plots on demand, add logic to randomly
select an available grid-aligned square region in a planting zone.

### CODE REVIEW

**Comment by sgrimm** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #1642** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1642?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
    * **PR #1644** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1644?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #1645** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1645?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/terraware/terraware-server/1642?utm_source=stack-comment).

**Review by nickgraz — APPROVED** [review]
Wow, this is awesome, nice work! One small comment but otherwise looks great.

File: src/main/kotlin/com/terraformation/backend/tracking/model/PlantingZoneModel.kt [hunk 1]
```diff
@@ -127,6 +137,174 @@ data class PlantingZoneModel(
...
+      val quadrants: MutableList<Pair<Polygon, Double>> =
+          listOf(
+                  rectangle(southwest.x, southwest.y, middleX, middleY),
+                  rectangle(middleX, southwest.y, northeast.x, middleY),
+                  rectangle(middleX, middleY, northeast.x, northeast.y),
+                  rectangle(southwest.x, middleY, middleX, northeast.y))
+              .map { quadrant -> quadrant to zoneGeometry.intersection(quadrant).area }
+              // Stop drilling down when we get to quadrants with 4 square meters of usable area
+              // since they're unlikely to round to different grid points than their parents.
+              .filter { it.second > 16.0 }
```
Comment (nickgraz) [inline src/main/kotlin/com/terraformation/backend/tracking/model/PlantingZoneModel.kt]: Can you explain this a bit more? 16 seems like a magic number

Comment (sgrimm) [reply src/main/kotlin/com/terraformation/backend/tracking/model/PlantingZoneModel.kt]: Good catch. This magic number made more sense in an earlier version of the code and I neglected to change it. Will address this.

Comment (sgrimm) [reply src/main/kotlin/com/terraformation/backend/tracking/model/PlantingZoneModel.kt]: Reworked this a bit so the base case of the recursion is at the top of the function. Hopefully it's clearer what's going on now.

**Comment by sgrimm** [comment]
Going to make this one a draft again. As part of the next PR in the series, I started trying this on actual maps, and the distortion is unacceptably high on very large regions.

**Comment by sgrimm** [comment]
This is ready for another look.

File: src/main/kotlin/com/terraformation/backend/tracking/model/PlantingZoneModel.kt [hunk 2]
```diff
@@ -190,12 +190,28 @@ data class PlantingZoneModel(
...
+     * should produce a true axis-aligned square at any distance from the origin.
      */
     fun gridAlignedSquare(west: Double, south: Double): Polygon {
-      val roundedWest = roundToGrid(west)
-      val roundedSouth = roundToGrid(south)
-      return rectangle(
-          roundedWest, roundedSouth, roundedWest + sizeMeters, roundedSouth + sizeMeters)
+      val start = geometryFactory.createPoint(Coordinate(roundToGrid(west), roundToGrid(south)))
+
+      return Turtle.makePolygon(start, meterCrs) {
```
Comment (nickgraz) [inline src/main/kotlin/com/terraformation/backend/tracking/model/PlantingZoneModel.kt:200]: 🚀

**Review by nickgraz — APPROVED** [review]
Looks great, nice work @sgrimm. Perhaps we should add one of those more complex maps as a test case?

**Comment by sgrimm** [comment]
### Merge activity

* **Jan 26, 1:17 PM**: @sgrimm started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/terraware/terraware-server/1642).
* **Jan 26, 1:17 PM**: @sgrimm merged this pull request with [Graphite](https://app.graphite.dev/github/pr/terraware/terraware-server/1642).

**Pull request merged**